File: java/src/main/java/br/com/emendes/removenthnodefromendoflist/RemoveNThNodeFromEndOfList.java
package br.com.emendes.removenthnodefromendoflist;

/**
 * Dado o cabeçalho de uma lista encadeada, remova o enésimo nó do final da lista e retorne seu cabeçalho.<br><br>
 * Restrições:<br>
 * 1 <= sz <= 30<br>
 * 0 <= Node.val <= 100<br>
 * 1 <= n <= sz
 */
public class RemoveNThNodeFromEndOfList {

  // Solução em 0ms.
  public ListNode removeNthFromEnd(ListNode head, int n) {
    ListNode actual = head;
    ListNode previous = head;
    int count = -1;
    while (count < n && actual != null) {
      actual = actual.next;
      count++;
    }

    if (actual == null && count < n) return head.next;

    while (actual != null) {
      actual = actual.next;
      previous = previous.next;
    }

    previous.next = previous.next.next;

    return head;
  }

  public static class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
      this.val = val;
    }

    ListNode(int val, ListNode next) {
      this.val = val;
      this.next = next;
    }

    int size() {
      int count = 0;
      ListNode actual = this;
      while (actual != null) {
        actual = actual.next;
        count++;
      }

      return count;
    }

    @Override
    public String toString() {
      if (this.next != null){
        return String.format("%d %s", this.val, next.toString());
      }
      return String.format("%d", this.val);
    }
  }

  public static void main(String[] args) {

    ListNode head = new ListNode(1);
    ListNode actual = head;

    for (int i = 2; i < 8; i++) {
      actual.next = new ListNode(i);
      actual = actual.next;
    }

    System.out.println(head);

    RemoveNThNodeFromEndOfList solution = new RemoveNThNodeFromEndOfList();
    ListNode newHead = solution.removeNthFromEnd(head, 1);

    System.out.println(newHead);

  }
}

File: javascript/problems/RemoveNthNodeFromEndOfList.js
/**
 * Dado o head de uma linked list, remova o nth node da fim da lista e retorne o head.
 *
 * Restrições:
 *
 * O número de nodes na lista is sz.
 *
 * 1 <= sz <= 30
 *
 * 0 <= Node.val <= 100
 *
 * 1 <= n <= sz
 *
 * Acompanhamento: Você conseguiria fazer isso de uma só vez?
 *
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  let dummy = new ListNode(-1, head);
  let count = 1;
  while (count < n) {
    head = head.next;
    count++;
  }
  let prev = dummy;
  let node = dummy;
  while (head != null) {
    head = head.next;
    prev = node;
    node = node.next;
  }
  prev.next = node.next;
  node.next = null;
  return dummy.next;
};

var firstSolution = function (head, n) {
  let dummy = new ListNode(-1, head);
  let size = 0;
  let node = head;
  while (node != null) {
    size++;
    node = node.next;
  }
  node = dummy;
  let previous = dummy;
  let count = 0;
  let nth = size - n + 1;
  while (count < nth) {
    previous = node;
    node = node.next;
    count++;
  }

  previous.next = node.next;
  node.next = null;
  return dummy.next;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}
